File: server/internal/usecase/web/dto/ticket.go
package dto

import "time"

type TicketListQuery struct {
	Page     int    `form:"page" validate:"omitempty,min=1"`
	PerPage  int    `form:"per_page" validate:"omitempty,min=1,max=100"`
	Status   string `form:"status" validate:"omitempty,oneof=waiting_admin waiting_user closed"`
	Category string `form:"category" validate:"omitempty,oneof=account order product technical billing other"`
	Priority string `form:"priority" validate:"omitempty,oneof=low normal high urgent"`
	OrderNo  string `form:"order_no" validate:"omitempty,max=64"`
}

type TicketCreateRequest struct {
	Title    string `validate:"required,max=160"`
	Category string `validate:"required,oneof=account order product technical billing other"`
	Priority string `validate:"omitempty,oneof=low normal high urgent"`
	Content  string `validate:"required,max=5000"`
	OrderNo  string `validate:"omitempty,max=64"`
}

type TicketMessageRequest struct {
	Content string `validate:"required,max=5000"`
}

type TicketCloseRequest struct {
	Reason *string `json:"reason" validate:"omitempty,max=500"`
}

type TicketItem struct {
	TicketNo      string     `json:"ticket_no"`
	Title         string     `json:"title"`
	Category      string     `json:"category"`
	Priority      string     `json:"priority"`
	Status        string     `json:"status"`
	OrderNo       *string    `json:"order_no"`
	LastMessageAt time.Time  `json:"last_message_at"`
	CreatedAt     time.Time  `json:"created_at"`
	ClosedAt      *time.Time `json:"closed_at"`
}

type TicketDetail struct {
	TicketItem
	CloseReason *string         `json:"close_reason"`
	Messages    []TicketMessage `json:"messages"`
}

type TicketMessage struct {
	ID          uint64             `json:"id"`
	SenderType  string             `json:"sender_type"`
	SenderName  string             `json:"sender_name"`
	Content     string             `json:"content"`
	Attachments []TicketAttachment `json:"attachments"`
	CreatedAt   time.Time          `json:"created_at"`
}

type TicketAttachment struct {
	FileID       uint64 `json:"file_id"`
	OriginalName string `json:"original_name"`
	MimeType     string `json:"mime_type"`
	Extension    string `json:"extension"`
	Size         uint64 `json:"size"`
	DownloadURL  string `json:"download_url"`
}

File: web/src/api/ticket.ts
import { request, type WebApiEnvelope } from './request'
import type { PageResponse } from './order'

export type TicketStatus = 'waiting_admin' | 'waiting_user' | 'closed'
export type TicketCategory = 'account' | 'order' | 'product' | 'technical' | 'billing' | 'other'
export type TicketPriority = 'low' | 'normal' | 'high' | 'urgent'
export type TicketSenderType = 'user' | 'admin'

export interface TicketTagItem {
  id: number
  name: string
  color: string | null
  visibility: 'public'
}

export interface TicketAttachment {
  file_id: number
  original_name: string
  mime_type: string
  extension: string
  size: number
  download_url: string
}

export interface TicketMessage {
  id: number
  sender_type: TicketSenderType
  sender_name: string
  content: string
  attachments: TicketAttachment[]
  created_at: string
}

export interface TicketItem {
  ticket_no: string
  title: string
  category: TicketCategory
  priority: TicketPriority
  status: TicketStatus
  tags: TicketTagItem[]
  order_no: string | null
  last_message_at: string
  created_at: string
  closed_at: string | null
}

export interface TicketDetail extends TicketItem {
  close_reason: string | null
  messages: TicketMessage[]
}

export interface TicketListQuery {
  page?: number
  per_page?: number
  status?: string
  category?: string
  priority?: string
  order_no?: string
}

export interface CreateTicketPayload {
  title: string
  category: string
  priority?: string
  content: string
  order_no?: string
  files?: File[]
}

export async function getTickets(params?: TicketListQuery) {
  const response = await request.get<WebApiEnvelope<PageResponse<TicketItem>>>('/tickets', { params })
  return response.data.data
}

export async function createTicket(payload: CreateTicketPayload) {
  const formData = new FormData()
  formData.append('title', payload.title)
  formData.append('category', payload.category)
  formData.append('priority', payload.priority || 'normal')
  formData.append('content', payload.content)
  if (payload.order_no) formData.append('order_no', payload.order_no)
  payload.files?.forEach((file) => formData.append('attachments', file))
  const response = await request.post<WebApiEnvelope<TicketDetail>>('/tickets', formData)
  return response.data.data
}

export async function getTicketDetail(ticketNo: string) {
  const response = await request.get<WebApiEnvelope<TicketDetail>>(`/tickets/${encodeURIComponent(ticketNo)}`)
  return response.data.data
}

export async function replyTicket(ticketNo: string, content: string, files: File[]) {
  const formData = new FormData()
  formData.append('content', content)
  files.forEach((file) => formData.append('attachments', file))
  const response = await request.post<WebApiEnvelope<TicketDetail>>(`/tickets/${encodeURIComponent(ticketNo)}/messages`, formData)
  return response.data.data
}

export async function closeTicket(ticketNo: string, reason?: string) {
  const response = await request.post<WebApiEnvelope<TicketDetail>>(`/tickets/${encodeURIComponent(ticketNo)}/close`, { reason })
  return response.data.data
}

export async function downloadTicketAttachment(ticketNo: string, fileId: number) {
  const response = await request.get<Blob>(`/tickets/${encodeURIComponent(ticketNo)}/attachments/${fileId}/download`, {
    responseType: 'blob',
  })
  return response.data
}
